File: java/src/main/java/io/github/alexmercedcoder/oap/Oap.java
package io.github.alexmercedcoder.oap;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import java.util.List;

/** Shared OAP constants and immutable result types. */
public final class Oap {
  /** OAP specification version implemented by this library. */
  public static final String SPEC_VERSION = "1.0";
  /** Java support-library version. */
  public static final String SUPPORT_VERSION = "1.0.4";
  private Oap() {}

  /** Supported document encoding. */
  public enum Format { YAML, JSON, MARKDOWN }
  /** One validation issue located by JSON Pointer. */
  public record Issue(String pointer, String message) {}
  /** Full-profile and immutable-specification identities. */
  public record Digests(String profile, String spec) {}
  /** Complete schema, semantic, and security validation result. */
  public record ValidationReport(String kind, ObjectNode document, List<Issue> errors, List<Issue> warnings, Digests digests, boolean ok) {}
  /** A requested policy value narrowed by an effective ceiling. */
  public record Adjustment(String field, JsonNode requested, JsonNode effective, String reason) {}
  /** Successful result of atomically applying a state delta. */
  public record DeltaApplication(ObjectNode profile, List<String> warnings, List<ObjectNode> pendingProposals) {}

  static ObjectNode object(JsonNode value) { return value != null && value.isObject() ? (ObjectNode) value : JsonSupport.emptyObject(); }
}

File: oap.go
// Package oap implements Open Agent Profile 1.0 parsing, validation,
// composition, policy narrowing, prompt rendering, and safe state deltas.
package oap

const (
	OAPVersion     = "1.0"
	SupportVersion = "1.0.5"
)

type Document map[string]any
type AgentProfile = Document
type AgentStateDelta = Document

type Issue struct {
	Pointer string `json:"pointer"`
	Message string `json:"message"`
}
type Digests struct {
	Profile string `json:"profile"`
	Spec    string `json:"spec"`
}
type Report struct {
	Kind     string   `json:"kind"`
	Document Document `json:"document,omitempty"`
	Errors   []Issue  `json:"errors"`
	Warnings []Issue  `json:"warnings"`
	Digests  *Digests `json:"digests,omitempty"`
	OK       bool     `json:"ok"`
}
type Adjustment struct {
	Field     string `json:"field"`
	Requested any    `json:"requested"`
	Effective any    `json:"effective"`
	Reason    string `json:"reason"`
}
type DeltaApplication struct {
	Profile          AgentProfile
	Warnings         []string
	PendingProposals []Document
}

func obj(value any) map[string]any {
	if v, ok := value.(map[string]any); ok {
		return v
	}
	return map[string]any{}
}
func arr(value any) []any {
	if v, ok := value.([]any); ok {
		return v
	}
	return nil
}
func strs(value any) []string {
	out := []string{}
	for _, v := range arr(value) {
		if s, ok := v.(string); ok {
			out = append(out, s)
		}
	}
	return out
}
func text(value any) string { s, _ := value.(string); return s }
